### PR TITLE
Deploy Independently Principle

### PR DESCRIPTION
**Pull request by domudall**
A proposal to add a principle of deploying microservices independently of each other.

* Add proposal.
* Update README with table of contents.

Please post all discussion notes either as comments directly on the Markdown, or within this pull request, as described in the [contributing guidelines](https://github.com/fresh8/microservice-principles/blob/master/CONTRIBUTING.md#discussing-changes).

### CODE REVIEW

File: 0002-deploy-independently.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+## Context
+
+In a microservice oriented system, being able to deploy independently allows for individual services to produce features and fix issues at an accelerated pace, due to a reduced overhead in release coordination. This also helps to provide a bounded context of a given service, and reduce coupling between others. All of these are key components of a continuous delivery system; anything that would require more advanced coordination can make even automation error-prone. Independent services are advantageous when moving to container systems such as Docker, as they are designed to be pared down to the minimal viable pieces needed to run whatever the one thing the container is designed to do, rather than packing multiple functions into the same machine.
+
+## Decision
+
+* One service per host; this host could be bare metal, a VM, or a container.
+* Service APIs should be robust:
+    * Use [tolerant readers](https://martinfowler.com/bliki/TolerantReader.html); only read elements you need from an API, and ignore the rest.
+    * When used, versioned APIs should co-exist to allow for deprication over time by other services.
```
Comment (nicolaslt) [inline 0002-deploy-independently.md]: Deprication -> deprecation

File: 0002-deploy-independently.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+Proposed
+
+## Context
+
+In a microservice oriented system, being able to deploy independently allows for individual services to produce features and fix issues at an accelerated pace, due to a reduced overhead in release coordination. This also helps to provide a bounded context of a given service, and reduce coupling between others. All of these are key components of a continuous delivery system; anything that would require more advanced coordination can make even automation error-prone. Independent services are advantageous when moving to container systems such as Docker, as they are designed to be pared down to the minimal viable pieces needed to run whatever the one thing the container is designed to do, rather than packing multiple functions into the same machine.
+
+## Decision
+
+* One service per host; this host could be bare metal, a VM, or a container.
```
Comment (nicolaslt) [inline 0002-deploy-independently.md]: I feel this is not related to Independent deployement. Not that I currently plan to but it wouldn't shock me to have more than one service per VM assuming they were reasonably scoped and completely interoperable.
Colocation can make sense to save latency for example.

File: 0002-deploy-independently.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+* Service APIs should be robust:
+    * Use [tolerant readers](https://martinfowler.com/bliki/TolerantReader.html); only read elements you need from an API, and ignore the rest.
+    * When used, versioned APIs should co-exist to allow for deprication over time by other services.
+* Services should be architectured for failure:
+    * Services should be tolerant of network/dependency failures, and retry connections with exponential backoff until a given cutoff time.
```
Comment (nicolaslt) [inline 0002-deploy-independently.md]: Should we add that services shouldn't assume a previously resolved IP will stay the same?

File: 0002-deploy-independently.md (hunk 1 above)
Comment (nicolaslt) [reply 0002-deploy-independently.md]: I would like this paragraph to become a "no-breaking changes" policy.
Running multiple APIs version is one way to achieve this but not always appropriate. The same can be achieved by hiding new behaviour behind parameters, or new url path, or other means.

**Comment by domudall** [comment]
Closing in favour of `Independence` principle.
